contracts/token/ERC20/ERC20.sol

```diff
@@ -299,13 +299,15 @@ abstract contract ERC20 is Context, IERC20, IERC20Metadata, IERC20Errors {
      * Does not emit an {Approval} event.
      */
     function _spendAllowance(address owner, address spender, uint256 value) internal virtual {
-        uint256 currentAllowance = allowance(owner, spender);
-        if (currentAllowance != type(uint256).max) {
-            if (currentAllowance < value) {
-                revert ERC20InsufficientAllowance(spender, currentAllowance, value);
-            }
-            unchecked {
-                _approve(owner, spender, currentAllowance - value, false);
+        if (value > 0) {
+            uint256 currentAllowance = allowance(owner, spender);
+            if (currentAllowance != type(uint256).max) {
+                if (currentAllowance < value) {
+                    revert ERC20InsufficientAllowance(spender, currentAllowance, value);
+                }
+                unchecked {
+                    _approve(owner, spender, currentAllowance - value, false);
+                }
             }
         }
```

Comment:
Didn't we decide against [address poisoning](https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3931) for similar reasons we would decide against filtering `value == 0`?

Reply:
The logic here is the following:
- if the temporary allowance is enough, we should not sload/sstore the persistent value (otherwize it break the point of gas savings). So we can do it in two ways:
  - in `ERC20TemporaryApproval._spendAllowance` only do the super call if value > 0
    - this is bad practice if someone else overrides `_spendAllowance`
  - in `ERC20` change the semantics of `_spendAllowance` to mean "if there is nothing to spend, we are good anyway".
  
If we look at `ERC20._spendAllowance`, this has the following impact
- if value = 0, `currentAllowance` cannot be smaller than value. `ERC20InsufficientAllowance` is never triggered, so the if doesn't change anything regarding the revert.
- if value is 0, 5.0 code does:
  - load the allowance (from a potentially overridable function)
  - substract zero from it
  - set it as the new allowance, without emitting an event.
  
So there is a change, we are no longer calling `_approve` with the current value. It may be possible to create edge cases where the missed call to an overrident `_approve` has an effect. IMO its a non issue.

Reply:
About address poisoning, it doesn't chagne the possibility of doing it (or not doing it). It makes it cheaper though, because the poisonning call would not read/write the zero allowance.

Reply:
@frangio curious to having your opinion on this if in the core ERC20.

Reply:
> It may be possible to create edge cases where the missed call to an overrident `_approve` has an effect.

Hm, I have a vague memory that this actually was a concern for a project once... It does seem quite risky to change this in a minor version.

Reply:
Could we implement this by overriding only `allowance` and `_approve` instead of `_spendAllowance`?

Reply:
> Could we implement this by overriding only allowance and _approve instead of _spendAllowance?

I have no idea how. In particular I'm not sure how to make transferFrom spend only temporary allowance, without touching the normal allowance, when the temporary allowance is enough

Reply:
Discussion on the ERC wording here: https://github.com/ethereum/ERCs/pull/358/commits/3c10621486d401040e187e56c31e570f292926c4#r1633595540

Reply:
I think it's possible, but not without the same problem of not invoking `super._approve`. There seems to be no way around that.

Reply:
> - in `ERC20TemporaryApproval._spendAllowance` only do the super call if value > 0
> 
>   -  this is bad practice if someone else overrides `_spendAllowance`

This is true, but it may be preferable than changing the behavior in a minor version. I think it's worth considering.

Reply:
Its what I used to do, before realizing there was a way to have the super call always happen.
I'm leanning toward the current version, but I'm open to re-using the old one

Reply:
As discussed, I re-implemented the old one.

It negativelly affect the tests because the error emitted in some edge conditions depends on _approve being called (with spender = 0 and value = 0).